From 4a76cf8c6ec1e61f196811b8a50d85d532dbe831 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Tue, 25 Jul 2017 08:03:07 -0400
Subject: [PATCH] memory leak in ReadOneMNGImage #619

A memory leak vulnerability was found in function ReadOneMNGImage,
which allow attackers to cause a denial of service (memory leak) via
a crafted file.

bug-debian: https://bugs.debian.org/870117
bug: https://github.com/ImageMagick/ImageMagick/issues/619
origin: https://github.com/ImageMagick/ImageMagick/commit/64e246252dff5bb16258adbc096da9bdc910bcd3
(cherry picked from commit 64e246252dff5bb16258adbc096da9bdc910bcd3)
---
 coders/png.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/coders/png.c b/coders/png.c
index b6a0cba45..deee821e6 100644
--- a/coders/png.c
+++ b/coders/png.c
@@ -5141,7 +5141,11 @@ static Image *ReadOneMNGImage(MngInfo* mng_info, const ImageInfo *image_info,
             mng_info->dhdr_warning++;
           }
         if (memcmp(type,mng_MEND,4) == 0)
-          break;
+          {
+            if (length != 0)
+              chunk=(unsigned char *) RelinquishMagickMemory(chunk);
+            break;
+          }
 
         if (skip_to_iend)
           {
